src/library/scala/collection/mutable/ArrayDeque.scala

```diff
@@ -248,15 +248,23 @@ class ArrayDeque[A] protected (
         reset(array = array2, start = 0, end = finalLength)
       } else if (2*idx <= finalLength) { // Cheaper to move the prefix right
         var i = suffixStart
+        while(i > removals) {
+          i -= 1
+          _set(i, _get(i - removals))
+        }
         while(i >= 0) {
           i -= 1
-          _set(i, if (i >= removals) _get(i - removals) else null.asInstanceOf[A])
+          _set(i, null.asInstanceOf[A])
```

Comment:
I don't understand this logic.  Why set -1 to `null.asInstanceOf[A]`?  I guess it probably doesn't hurt since there has to be a gap there (one element isn't used).

Reply:
(I'm thinking about the case where `i == 0` starting out.)

Reply:
yeah it was confusing to me too.  For simple cases it does one additional null set, but I think it's on purpose to handle nulling out a value that moves

It's undocumented and not well tested -- I had a typo in an earlier version and the tests still passed...

Reply:
That's what I thought at first, but `i == -1` means _one element before the start_, which was presumably null to begin with, and shouldn't have been touched by the remove operation.  So it looks to me like probably an off-by-one error, though not one that matters because there's a dead element anyway?  Or maybe it was copied from some other context where it _was_ important to write a null element before start?

Reply:
(i'm also _slighly_ worried by not being confident about what's going on here in light of your other change which reduces the extra space by one cell.  I _think_ it's still fine, and I _also_ think this is just a bug, but my confidence is middling at best because there is so much detailed indexing in this implementation (as is typically necessary for decent performance in a ring buffer).)

Reply:
i == -1 wraps around and sets the null sentinel at the end to null

Reply:
Yeah I guess I wasn’t 100% sure but I think it’s on purpose. The intention in this PR is to not change any semantics, so that only a benchmark is needed.

Reply:
I've updated the patch to fix this as well.

Reply:
@isaacl Since off-by-one errors are always worrying, are there any unit tests covering the cases in which an `ArrayDequeue` is full and up-scaled, in which the ends are both in the beginning or the end of the array?

Reply:
not checked in but the benchmark tests that